File: python/azure/devopsagents.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from __future__ import (division, absolute_import, print_function,
                        unicode_literals)

from dataclasses import dataclass, field
import base64
import json
import logging
import os
import sys
from waitress import serve
import requests
from flask import Flask


@dataclass
class RunJob:

    logging.basicConfig(
        format="%(asctime)s %(levelname)s: %(message)s",
        level=logging.INFO,
        datefmt="%d-%b-%y %H:%M:%S",
        stream=sys.stderr,
    )

    auth: str = field(default="")
    encodeauth: str = field(default="")
    headers: dict = field(default_factory=dict)
    secret: str = os.getenv("ADV_TOKEN", "")
    organization: str = os.getenv("ADV_ORG", "inc")
    poolid: str = os.getenv("ADV_POOL", "1")

    def __post_init__(self):
        self.authencode()

    def authencode(self):

        self.auth = f':{self.secret}'
        self.encodeauth = base64.b64encode(self.auth.encode('utf-8'))

        self.headers = {
            'Authorization':
            'Basic {0}'.format(self.encodeauth.decode('utf-8'))
        }

    def get_running(self):

        url = (
            f"https://dev.azure.com/{self.organization}/"
            f"_apis/distributedtask/pools/{self.poolid}/jobrequests")

        try:
            response = requests.request("GET", url, headers=self.headers)
        except NameError as e:
            logging.error(e)

        if response.ok:
            try:
                data = json.loads(response.text).get("value")
                results = self.get_data(data)
                return {'runningjobs': results}
            except json.JSONDecodeError as e:
                logging.error(e)

        return {'runningjobs': ""}

    def get_data(self, data):
        results = len([item for item in data if "result" not in item])
        return results


app = Flask(__name__)
jobrequests = RunJob()


@app.route('/')
def home():
    print(jobrequests.get_running(), file=sys.stderr)
    return jobrequests.get_running()


def main():
    try:
        serve(app)
    except OSError as e:
        logging.error(e)


if __name__ == '__main__':
    main()

File: golang/azure/devopsagents.go
///bin/true; exec /usr/bin/env go run "" "$@"
package main

import (
	"encoding/base64"
	"net/http"
	"os"
)

var (
	auth         string
	encodeauth   string
	headers      map[string]string
	secret       string
	organization string
	poolid       string
)

// "github.com/Azure/go-autorest/autorest/azure/auth"

func authencode() string {

	token, tokenpresent := os.LookupEnv("ADV_TOKEN")

	if tokenpresent {
		auth = ":" + token
		encodeauth = base64.StdEncoding.EncodeToString([]byte(auth))
	}
	return encodeauth

}

func get_running() {

	org, _ := os.LookupEnv("ADV_ORG")
	pool, _ := os.LookupEnv("ADV_POOL")

	url := "https://dev.azure.com/" + org + "/_apis/distributedtask/pools/" + pool + "/jobrequests"

	request, _ := http.NewRequest("GET", url, nil)
	request.Header.Add("Authorization", "Basic "+encodeauth)
}

func main() {

	authencode()

	// authorizer, err := auth.NewAuthorizerFromEnvironment()

}
